Nodejs week1/farzaneh

FarzanehAhmadi opened the pull request:
Hi!
I converted the meal table to an array for now since we haven’t learned how to use databases with Node.js yet. Just trying to get the routes working!
I also hardcoded the server port because the .env file wasn’t working, even though I created it and copied .env-example to that.
Thanks a lot for reviewing this! 🙏

FarzanehAhmadi committed:
Add routes for meals

FarzanehAhmadi committed:
Commented out unused routes and server setup

FarzanehAhmadi committed:
Setup server and uncomment app.use("/api", apiRouter)

FarzanehAhmadi committed:
Used knex.raw() for getting meals from the DB instead of hardcoded array

FarzanehAhmadi committed:
Change response from send() to json()

FarzanehAhmadi committed:
Polish

FarzanehAhmadi committed:
Update SQL queries

FarzanehAhmadi committed:
Polish

FarzanehAhmadi committed:
Connect database

FarzanehAhmadi committed:
Woohoo, it's running on port 3001! 🎉

kiko-bkr commented on api/src/database_client.js:
I see that you did not create a `.env` file, and that is why you had to enter the values into the code here. You should make a copy of the `.env-example` file, and rename it to `.env`, and afterwards replace the values with your database's values.

FarzanehAhmadi replied:
I have created a .env file, but since it's in .gitignore, it doesn't appear here. However, it exists locally and works on port 3001. Should I remove .env from .gitignore, or is keeping it ignored the best practice?

kiko-bkr replied:
No, the `.env` file should never be pushed to version control. (for example Git) That file should only exist on your own machine, since it contains your sensitive information, such as database passwords.

kiko-bkr commented on api/src/database_client.js:
```diff
@@ -1,16 +1,41 @@
+
+export default db;
+
+// import knex from "knex";
+
+// const connection = knex({
+//   client: process.env.DB_CLIENT,
+//   connection: {
+//     host: process.env.DB_HOST,
+//     port: process.env.DB_PORT,
+//     user: process.env.DB_USER,
+//     password: process.env.DB_PASSWORD,
+//     database: process.env.DB_DATABASE_NAME,
+//     ssl:
+//       process.env.DB_USE_SSL === "true" ? { rejectUnauthorized: false } : false,
+//   },
+// });
+
+// export default connection;
```
If you create a `.env` file, this code will work.

kiko-bkr commented on api/src/index.js:
```diff
@@ -11,21 +11,97 @@ app.use(bodyParser.json());
+apiRouter.get("/future-meals", async (req, res) => {
+  try {
+    const futureMeals = await knex.raw("Select * From Meal WHERE Meal.when > NOW()")
+    if (futureMeals[0].length === 0) {
+    return res.status(404).json("There are no meals for the future.");
+    }
```
In such cases you should not return a `Not Found` error. If there are no meals, and the user wants to see all of the future ones, then returning an empty array is a valid answer.

kiko-bkr commented on api/src/index.js:
```diff
@@ -11,21 +11,97 @@ app.use(bodyParser.json());
+apiRouter.get("/past-meals", async (req, res) => {
+  try {
+    const pastMeals = await knex.raw("Select * From Meal WHERE Meal.when < NOW()") 
+  if(pastMeals[0].length === 0) {
+    return res.status(404).json("There are no meals in the past.")
+  }
```
Same here.

kiko-bkr commented on api/src/index.js:
```diff
@@ -11,21 +11,97 @@ app.use(bodyParser.json());
+apiRouter.get("/first-meal", async (req, res)=>{
+  try {
+    const firstMeal = await knex.raw("SELECT * FROM Meal ORDER BY id ASC LIMIT 1");
+    if(firstMeal[0].length===0){
+      return res.status(404).json("No meals found!")
+    }
```
Here it does make sense to have a `404` error! Good job! 👏  

To make it clear: returning a `404` error makes sense in endpoints where the user expects a specific meal to be returned. Another example would be when the user would want to query a meal with a specific ID: if it does not exist, we should return `404`.

kiko-bkr commented on api/src/index.js:
```diff
@@ -11,21 +11,97 @@ app.use(bodyParser.json());
+    if(firstMeal[0].length===0){
+      return res.status(404).json("No meals found!")
+    }
+    res.json(firstMeal[0][0])
```
A small suggestion is to use some variables to make the code more readable. Looking at `firstMeal[0][0]` is a bit confusing. You can make this easier to understand by having something like:
```javascript
const result = await knex.raw(...);
const meals = result[0];
const firstMeal = meals[0];
```

kiko-bkr commented on api/src/index.js:
```diff
@@ -11,21 +11,97 @@ app.use(bodyParser.json());
+    if(lastMeal[0].length === 0){
+      return res.status(404).json("No meals found!")
+    }
+    res.json(lastMeal[0][0])
```
Same applies here.

kiko-bkr reviewed (changes requested):
Well implemented! Your code meets all the requirements of the assignment. 

Two main comments:

1.  Make sure to read your database configuration from an `.env` file. I left instruction on how to do that.
2. Do not return 404 for endpoints that are supposed to return multiple objects.

Address these comments, and the code will be great! ⚡

FarzanehAhmadi committed:
Return empty array instead of 404 for no future or past meals

FarzanehAhmadi committed:
Refactor /first-meal and /last-meal for readability and error handling

FarzanehAhmadi commented:
Hi, thanks for your feedback! I really appreciate it.🌿

I’ve addressed comments regarding the 404 response for endpoints returning multiple objects. Regarding the .env file, I have already created it, but since it's in .gitignore (as per the template), it doesn’t show up here. Let me know if it's necessary to remove it from .gitignore, or if keeping it local is fine.

FarzanehAhmadi committed:
Revert file to initial state

kiko-bkr commented on api/src/index.js:
```diff
@@ -11,21 +11,98 @@ app.use(bodyParser.json());
+apiRouter.get("/future-meals", async (req, res) => {
+  try {
+    const futureMeals = await knex.raw("Select * From Meal WHERE Meal.when > NOW()")
+    if (futureMeals[0].length === 0) {
+    return res.json([]);
+    }
+    res.json(futureMeals[0]);
```
This is just a small sidenote, but if `futureMeals[0].length === 0`, then that means that `futureMeals === []` anyway. 😄  This means that this `if` is not necessary, since once you are inside the `if`, the value of `futureMeals` is `[]` anyways. You can just return `futureMeals` without checking for length. 👍

kiko-bkr commented on api/src/index.js:
```diff
@@ -11,21 +11,98 @@ app.use(bodyParser.json());
+    if(meals[0].length === 0) {
+      return res.status(404).json('There are no meals!')
+    }
+    res.json(meals[0]);
```
Yes, here you do it exactly as you should 👏

kiko-bkr commented on api/src/index.js:
```diff
@@ -11,21 +11,98 @@ app.use(bodyParser.json());
+apiRouter.get("/first-meal", async (req, res)=>{
+  try {
+    const result = await knex.raw("SELECT * FROM Meal ORDER BY id ASC LIMIT 1");
+    const meals = result[0];
+    const firstMeal = meals[0];
+    if(!firstMeal){
+      return res.status(404).json("No meals found!")
+    }
+    res.json(firstMeal)
```
Very nice!

kiko-bkr reviewed (approved):
Well implemented! Left a small note, but the code functions just as expected! 👏
